CRM/Civicase/Page/ActivityFiles.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+class CRM_Civicase_Page_ActivityFiles {
+
+  /**
+   * Download all activity files contained in a single zip file.
+   */
+  public static function downloadAll() {
+    $activity = self::getActivityFromRequest();
+
+    $zipName = self::getZipName($activity);
+    $zipDestination = self::getDestinationPath();
+    $zipFullPath = $zipDestination . '/' . $zipName;
+    $files = self::getActivityFilePaths($activity['id']);
+    $zipFileResource = self::createZipFile($zipFullPath, $files);
+
+    unlink($zipFullPath);
+    self::downloadZipFileResource($zipName, $zipFileResource);
+  }
+
+  /**
+   * Returns the activity specified by the request. In case the request gives
+   * an invalid activity id it throws a 404 status code.
+   */
+  private static function getActivityFromRequest() {
+    $activityId = CRM_Utils_Array::value('activity_id', $_GET);
+
+    self::validateActivityId($activityId);
+
+    $activityResult = civicrm_api3('Activity', 'get', [
+      'id' => $activityId,
+      'return' => [ 'activity_type_id.label' ]
+    ]);
+
+    if ($activityResult['count'] === 0) {
+      return self::throwStatusCode('404 Not Found');
+    }
+
+    return CRM_Utils_Array::first($activityResult['values']);
+  }
+
+  /**
+   * Validates that the activity id was provided. If not, it returns a 404 status code.
+   *
+   * @param string|null $activityId
+   */
+  private static function validateActivityId($activityId) {
+    if (empty($activityId)) {
+      self::throwStatusCode('400 Bad Request');
+    }
+  }
+
+  /**
+   * Throws a specific status code and closes the connection.
+   *
+   * @param string $statusCode
+   */
+  private static function throwStatusCode($statusCode) {
+    CRM_Utils_System::setHttpHeader('Status', $statusCode);
+    CRM_Utils_System::civiExit();
+  }
+
+  /**
+   * Given an activity, it returns the name for the zip file containing all of
+   * its files. Ex: Activity Open Case 123.zip
+   *
+   * @param array $activity
+   *
+   * @return string
+   */
+  private static function getZipName($activity) {
+    return 'Activity ' . $activity['activity_type_id.label'] . ' ' . $activity['id'] . '.zip';
```

Comment:
Can we check what happens if activity label contains `"` or `'`?

Reply:
Here are some tests done after refactoring using `CRM_Utils_String::munge`.

input ---> output
Activity "Open Case" 1758.zip -> Activity Open Case 1758.zip
Activity 'Open Case' 1758.zip -> Activity Open Case 1758.zip
Activity \Open Case\ 1758.zip -> Activity Open Case 1758.zip
Activity /Open Case/ 1758.zip -> Activity Open Case 1758.zip